### ISSUE TITLE
[Bug?]: Dynamic import of module with query string param doesn't work

### ISSUE THREAD

**Issue by kubijo** [yarnpkg/berry#3894]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

When I enable the `pnpEnableEsmLoader` flag & try to import local module in this manner:

```javascript
import('./module.js?cache=busted');
```

I get errors such as these (real path prefixes replaced):

```
/path/to/.pnp.cjs:46662
  return Object.defineProperties(new Error(message), {
                                 ^

Error: Qualified path resolution failed - none of those files can be found on the disk.

Source path: /path/to/module.js?cache=busted
Not found: /path/to/module.js?cache=busted

    at internalTools_makeError (/path/to/.pnp.cjs:46662:34)
    at resolveUnqualified (/path/to/.pnp.cjs:48025:13)
    at resolveRequest (/path/to/.pnp.cjs:48058:14)
    at Object.resolveRequest (/path/to/.pnp.cjs:48130:26)
    at resolve$1 (file:///path/to/.pnp.loader.mjs:206:25)
    at ESMLoader.resolve (node:internal/modules/esm/loader:422:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:222:40)
    at ESMLoader.import (node:internal/modules/esm/loader:276:22)
    at importModuleDynamically (node:internal/modules/esm/translators:111:35)
    at importModuleDynamicallyCallback (node:internal/process/esm_loader:35:14)
```

-------------------------------------------------------------

The whole reason behind doing this is that [described here](https://ar.al/2021/02/22/cache-busting-in-node.js-dynamic-esm-imports/) (tl;dr: CommonJS allows you to manipulate its import cache, but ES-modules don't have that any more)

### To reproduce

**Minimal repro** is a rather esoteric concept at this stage & I do believe that this is a question of such imports even being supported by yarn.

If it is indeed supported, and you believe that this should work as is, I will gladly supply more information as requested.

### Environment

```shell
System:
    OS: Linux 5.11 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (12) x64 Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
  Binaries:
    Node: 16.13.1 - /tmp/xfs-3cc39734/node
    Yarn: 3.2.0-rc.5 - /tmp/xfs-3cc39734/yarn
```

**Comment by merceyz** [comment]
This is indeed a bug, fixed in https://github.com/yarnpkg/berry/pull/3895

You can try it out by running the following
```
yarn set version from sources --branch 3895
```

**Comment by kubijo** [comment]
Thank you, will try that tomorrow.
Any idea about a release our at least a merge into master?
Also do I understand it correctly that absolute paths are a different case and shouldn't work there?

**Comment by merceyz** [comment]
It has to be merged before we create a release, I don't know when it will be merged since someone other than myself needs to review it.

Absolute paths with search parameters already work, I added a test in https://github.com/yarnpkg/berry/pull/3895 to ensure it doesn't regress in the future.

**Comment by kubijo** [comment]
I see... I will to try if newer version is available and it works there because absolute path was my first case and it didn't work ... only then I tried changing it to relative.

**Comment by kubijo** [comment]
So I've tried it with a fresh canary and I get much the same.
Any idea if I could be screwing this up myself somehow?


```
  System:
    OS: Linux 5.11 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (12) x64 Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
  Binaries:
    Node: 16.13.1 - /tmp/xfs-3ab2aa32/node
    Yarn: 3.2.0-rc.8 - /tmp/xfs-3ab2aa32/yarn
```

```
/opt/dev/…/.pnp.cjs:46662
  return Object.defineProperties(new Error(message), {
                                 ^

Error: Qualified path resolution failed - none of those files can be found on the disk.

Source path: /opt/dev/…/webpack/webpack.config.gates.js?cache=busted
Not found: /opt/dev/…/webpack/webpack.config.gates.js?cache=busted

    at internalTools_makeError (/opt/dev/…/.pnp.cjs:46662:34)
    at resolveUnqualified (/opt/dev/…/.pnp.cjs:48025:13)
    at resolveRequest (/opt/dev/…/.pnp.cjs:48058:14)
    at Object.resolveRequest (/opt/dev/…/.pnp.cjs:48130:26)
    at resolve$1 (file:///opt/dev/…/.pnp.loader.mjs:206:25)
    at ESMLoader.resolve (node:internal/modules/esm/loader:422:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:222:40)
    at ESMLoader.import (node:internal/modules/esm/loader:276:22)
    at importModuleDynamically (node:internal/modules/esm/translators:111:35)
    at importModuleDynamicallyCallback (node:internal/process/esm_loader:35:14)
```

**Comment by kubijo** [comment]
OK, so I've tried your version @merceyz and sadly to no luck.
Also, I have got another observation… even if I use a relative path in the `import` function, the error message is always for absolute path…

```javascript
{
  configFilePath: '/opt/dev/…/conf/webpack/webpack.config.gates.js',braiins/pool/web-ui/conf
  configFilePathRelative: './webpack.config.gates.js?cache=busted',
}
```

```
Source path: /opt/dev/…/webpack/webpack.config.gates.js?cache=busted
Not found: /opt/dev/…/webpack/webpack.config.gates.js?cache=busted
```

**Comment by merceyz** [comment]
I'm going to need a reproduction then, based on the code you provided in the OP it works with #3895

**Comment by kubijo** [comment]
I believe I've able to set up something here: https://github.com/kubijo/yarn-es-import-query.

The idea is to run these to get a passing run of the test script:
- `yarn install`
- `yarn setup`
- `yarn test`

… and to test the failing imports, uncomment either pair of the cache-busting imports (I've added both relative and absolute examples).

It presumes though that you have `python3` with `venv` module in your system to manage `node` that it runs under through [nodeenv](https://pypi.org/project/nodeenv/).

… I'm confident, though, that If you don't, you'll manage to set it up according to your options.

**Comment by merceyz** [comment]
I see that the cache isn't busted so that is a bug but it does resolve correctly

**Comment by kubijo** [comment]
Oh... I don't get it resolved though! 🤔. I get the same error I described above

**Comment by merceyz** [comment]
To clarify: I see the same error as you with the version you checked in but not with https://github.com/yarnpkg/berry/pull/3895

**Comment by kubijo** [comment]
Ok, it seems we do get different behaviour thought... I will create a screen cast to squash any chance of misunderstanding when I get back home ... (won't be sooner then 3 hours from now)

**Comment by merceyz** [comment]
Fixed cache busting in https://github.com/yarnpkg/berry/pull/3895/commits/0594e33342b65e8e07538299f1b6e2d823a1c15c

**Comment by kubijo** [comment]
https://user-images.githubusercontent.com/11244314/147137122-300d8fc5-3f44-499d-8b83-b64138f67e4a.mp4

**Comment by kubijo** [comment]
If you have any idea how to progress with debugging, I am here and happy to help!

**Comment by kubijo** [comment]
Additionally… it is interesting to see that the node warning about loader being an experimental feature is still there even though master contains a commit that claims to have supressed it… although that is admitedly beside the topic at hand.

**Comment by merceyz** [comment]
Did you run `yarn install` in that folder? I see the version is correct based on the output of `yarn --version` but it doesn't look like you've run an install recently.

**Comment by kubijo** [comment]
Damn man!
I didn't after the version upgrade & it didn't seem to me that it should matter, so when you wrote this I was fully expecting to still fail, but it now certainly works ... both absolute and relative :open_mouth:
